doc/neps/nep-0042-new-dtypes.rst

```diff
@@ -0,0 +1,796 @@
+Such an array may include a mix of datatypes, as long as they can be clearly
+promoted.
+Currently not clearly distinct (but partially existing for strings) is the
+use case of::
+
+    # currently `np.String` would be spelled `"S"`. 
+    np.array([object(), None], dtype=np.String)
+
+which forces each object to be interpreted as string. This is anticipated
+to be useful for example for categorical datatypes::
+
+    np.array([1, 2, 1, 1, 2], dtype=Categorical)
+
+can be implemented.
+(For NumPy ``datetime64`` this is also currently used to allow string input.)
+
+There are three further issues to consider:
+
+1. It is desirable that datatypes can be created which are associated to normal
+   python scalars (such as ``datetime.Datetime``), which do not have a ``dtype``
+   attribute already.
+2. A general datatype could represent a sequence, however, NumPy currently
+   assumes that sequences always include elements (the sequence cannot be an
+   element itself). An example for this is a ``vector`` DType.
+3. An array may contain arrays with specific a dtype (even general python objects).
+   In the above example of ``np.array([np.array(None, dtype=object)], dtype=np.String)``
+   this creates the issue of how to handle the included array.
+
+Some of these difficulties arise due to the fact that finding the correct shape
+of the output array and finding the correct datatype are closely related.
+
+**Implementation:**
+There are two distinct cases given above: First, when the user has provided no
+dtype information, and second when the user provided a DType class – 
+a notion that is currently represented e.g. by ``"S"``, but not cleanly supported.
+
+In the first case, it is necessary to establish a mapping from the python type
+that was put in to the DType class.
+Further, the correct dtype instance needs to be found.
+This shall be implemented by the use of two informations:
+
+1. ``type``: The current type attribute to indicate which python type is
+   associated with the DType class (this is a *class* attribute that always
+   exist for any datatype and is not limited to array coercion).
+2. A flag allowing to to optionally not allow automatic discovery of the DType.
+   This allows e.g. the addition of a ``tuple`` dtype without changing the
+   parsing logic.   
+3. ``__discover_descr_from_pyobject__(cls, obj) -> dtype``: A classmethod that
+   returns the correct descriptor given the input object.
+   *Note that only flexible DTypes have to implement this*.
+
+The ``type`` which is already associated with any dtype through the ``dtype.type``
+attribute is associated to each class.
+This will be cached globally to create a mapping (dictionary)
+``knwon_python_types[type] = DType``.
+NumPy currently uses a small hard-coded a mapping and conversion of numpy scalars
+(inheriting from ``np.generic``) to achieve this, however, this forces a certain
+structure on the associated ``type``.
```

Comment:
Right, Python integers would probably map to -> `AbstractPyInt[value]` (a class), this can then return an actual dtype when asked for it (i.e. integer). At the same time, `common_dtype` can be defined correctly for it *and* whatever machinery we use for UFuncs has the information available needed for value based casting. Partially, I thought I can defer it to the UFuncs, but for array coercion it is also important.